fmriprep/config.py

```diff
@@ -1,5 +1,591 @@
+
+This module implements the memory structures to keep a consistent, singleton config.
+Settings are passed across processes via filesystem, and a copy of the settings for
+each run and subject is left under
+``<output_dir>/sub-<participant_id>/log/<run_unique_id>/fmriprep.toml``.
+Settings are stored using :abbr:`ToML (Tom's Markup Language)`.
+The module has a :py:func:`~fmriprep.config.to_filename` function to allow writting out
+the settings to hard disk in *ToML* format, which looks like:
+
+.. literalinclude:: ../fmriprep/data/tests/config.toml
+   :language: toml
+   :name: fmriprep.toml
+   :caption: **Example file representation of fMRIPrep settings**.
+
+This config file is used to pass the settings across processes,
+using the :py:func:`~fmriprep.config.load` function.
+
+Configuration sections
+----------------------
+.. autoclass:: environment
+   :members:
+.. autoclass:: execution
+   :members:
+.. autoclass:: workflow
+   :members:
+.. autoclass:: nipype
+   :members:
+
+Usage
+-----
+A config file is used to pass settings and collect information as the execution
+graph is built across processes.
+
+.. code-block:: Python
+
+    from fmriprep import config
+    config_file = config.execution.work_dir / '.fmriprep.toml'
+    config.to_filename(config_file)
+    # Call build_workflow(config_file, retval) in a subprocess
+    with Manager() as mgr:
+        from .workflow import build_workflow
+        retval = mgr.dict()
+        p = Process(target=build_workflow, args=(str(config_file), retval))
+        p.start()
+        p.join()
+    config.load(config_file)
+    # Access configs from any code section as:
+    value = config.section.setting
+
+Logging
+-------
+.. autoclass:: loggers
+   :members:
+
+Other responsibilities
+----------------------
+The :py:mod:`config` is responsible for other conveniency actions.
+
+  * Switching Python's :obj:`multiprocessing` to *forkserver* mode.
+  * Set up a filter for warnings as early as possible.
+  * Automated I/O magic operations. Some conversions need to happen in the
+    store/load processes (e.g., from/to :obj:`~pathlib.Path` \<-\> :obj:`str`,
+    :py:class:`~bids.layout.BIDSLayout`, etc.)
+
+"""
+from multiprocessing import set_start_method
+import warnings
+
+# cmp is not used by fmriprep, so ignore nipype-generated warnings
+warnings.filterwarnings('ignore', 'cmp not installed')
+warnings.filterwarnings('ignore', 'This has not been fully tested. Please report any failures.')
+warnings.filterwarnings('ignore', "sklearn.externals.joblib is deprecated in 0.21")
+warnings.filterwarnings('ignore', "can't resolve package from __spec__ or __package__")
+warnings.filterwarnings('ignore', category=DeprecationWarning)
+warnings.filterwarnings('ignore', category=FutureWarning)
+warnings.filterwarnings('ignore', category=ResourceWarning)
+
+
+try:
+    set_start_method('forkserver')
+except RuntimeError:
+    pass  # context has been already set
+finally:
```

Comment:
Theoretical reason: it is explicit those are to be run regardless the branch flow went through.
Practical reason: so that my linter does not complain about imports after plain code.

Happy to take out from the finally if that feels more appropriate.